### ISSUE TITLE
请问FDloss是怎么用在模型的损失函数中的？ 谢谢

### ISSUE THREAD

**Issue by 13088992520** [Kent0n-Li/GT-U-Net#3]

**Comment by Kent0n-Li** [comment]
FDloss is optional for our backbone network. If you want to use it, you can transfer the segmentation result and label to uint8 Numpy array, and send them to FDloss to calculate the Fourier descriptor difference.

**Comment by 13088992520** [comment]
Thankyou

**Comment by 13088992520** [comment]
还有一个问题没太搞清楚，如果把它放在损失函数中的话，那Fourier descriptor difference是会随训练变化的么， 那是把损失函数中的y_true 和y_pred 转化为numpy array再计算Fourier descriptor difference么，谢谢

**Comment by Kent0n-Li** [comment]
> 还有一个问题没太搞清楚，如果把它放在损失函数中的话，那Fourier descriptor difference是会随训练变化的么， 那是把损失函数中的y_true 和y_pred 转化为numpy array再计算Fourier descriptor difference么，谢谢

yes

**Comment by 13088992520** [comment]
Thank you so much

**Comment by junyuchen245** [comment]
> FDloss is optional for our backbone network. If you want to use it, you can transfer the segmentation result and label to uint8 Numpy array, and send them to FDloss to calculate the Fourier descriptor difference.

Hi @Kent0n-Li ,

I'm not sure how this is even feasible during network training. Both Pytorch and TF prohibit the use of numpy functions to calculate loss since this will detach the computation graph and make it impossible to use autograd. I checked your FDloss script and saw that you did not seem to have implemented a custom backprop function. Therefore, I'm wondering as to how you went about using FDloss for training. Could you provide further information about this?

Junyu

**Comment by Kent0n-Li** [comment]
> > FDloss is optional for our backbone network. If you want to use it, you can transfer the segmentation result and label to uint8 Numpy array, and send them to FDloss to calculate the Fourier descriptor difference.
> 
> Hi @Kent0n-Li ,
> 
> I'm not sure how this is even feasible during network training. Both Pytorch and TF prohibit the use of numpy functions to calculate loss since this will detach the computation graph and make it impossible to use autograd. I checked your FDloss script and saw that you did not seem to have implemented a custom backprop function. Therefore, I'm wondering as to how you went about using FDloss for training. Could you provide further information about this?
> 
> Junyu

You can read this paper https://arxiv.org/pdf/1904.10030.pdf

**Comment by junyuchen245** [comment]
> > > FDloss is optional for our backbone network. If you want to use it, you can transfer the segmentation result and label to uint8 Numpy array, and send them to FDloss to calculate the Fourier descriptor difference.
> > 
> > 
> > Hi @Kent0n-Li ,
> > I'm not sure how this is even feasible during network training. Both Pytorch and TF prohibit the use of numpy functions to calculate loss since this will detach the computation graph and make it impossible to use autograd. I checked your FDloss script and saw that you did not seem to have implemented a custom backprop function. Therefore, I'm wondering as to how you went about using FDloss for training. Could you provide further information about this?
> > Junyu
> 
> You can read this paper https://arxiv.org/pdf/1904.10030.pdf

I am fully aware of the implementation of HD loss. Due to the fact that the distance map used in the HD loss was considered to be constant during training (it is computed prior to training), autograd was achievable (see: https://github.com/JunMa11/SegLoss/issues/36#issuecomment-947804005). However, it seemed that the FD described in your paper was not constant during training. Could you elaborate on your loss and training strategies?

**Comment by Kent0n-Li** [comment]
> > > > FDloss is optional for our backbone network. If you want to use it, you can transfer the segmentation result and label to uint8 Numpy array, and send them to FDloss to calculate the Fourier descriptor difference.
> > > 
> > > 
> > > Hi @Kent0n-Li ,
> > > I'm not sure how this is even feasible during network training. Both Pytorch and TF prohibit the use of numpy functions to calculate loss since this will detach the computation graph and make it impossible to use autograd. I checked your FDloss script and saw that you did not seem to have implemented a custom backprop function. Therefore, I'm wondering as to how you went about using FDloss for training. Could you provide further information about this?
> > > Junyu
> > 
> > 
> > You can read this paper https://arxiv.org/pdf/1904.10030.pdf
> 
> I am fully aware of the implementation of HD loss. Due to the fact that the distance map used in the HD loss was considered to be constant during training (it is computed prior to training), autograd was achievable (see: [JunMa11/SegLoss#36 (comment)](https://github.com/JunMa11/SegLoss/issues/36#issuecomment-947804005)). However, it seemed that the FD described in your paper was not constant during training. Could you elaborate on your loss and training strategies?

You can try to adopt the same training strategy as HD loss. In the calculation of this article, we replaced the calculation of numpy with the implementation of pytorch. In our other subsequent projects, we tried the strategy of HD loss, which seems to work better.

**Comment by junyuchen245** [comment]
> > > > > FDloss is optional for our backbone network. If you want to use it, you can transfer the segmentation result and label to uint8 Numpy array, and send them to FDloss to calculate the Fourier descriptor difference.
> 
> > > > 
> 
> > > > 
> 
> > > > Hi @Kent0n-Li ,
> 
> > > > I'm not sure how this is even feasible during network training. Both Pytorch and TF prohibit the use of numpy functions to calculate loss since this will detach the computation graph and make it impossible to use autograd. I checked your FDloss script and saw that you did not seem to have implemented a custom backprop function. Therefore, I'm wondering as to how you went about using FDloss for training. Could you provide further information about this?
> 
> > > > Junyu
> 
> > > 
> 
> > > 
> 
> > > You can read this paper https://arxiv.org/pdf/1904.10030.pdf
> 
> > 
> 
> > I am fully aware of the implementation of HD loss. Due to the fact that the distance map used in the HD loss was considered to be constant during training (it is computed prior to training), autograd was achievable (see: [JunMa11/SegLoss#36 (comment)](https://github.com/JunMa11/SegLoss/issues/36#issuecomment-947804005)). However, it seemed that the FD described in your paper was not constant during training. Could you elaborate on your loss and training strategies?
> 
> 
> 
> You can try to adopt the same training strategy as HD loss. In the calculation of this article, we replaced the calculation of numpy with the implementation of pytorch. In our other subsequent projects, we tried the strategy of HD loss, which seems to work better.

I see. Thanks for the clarification.

**Comment by Kent0n-Li** [comment]
solved
